ensembl_genes/Red.py

```diff
@@ -0,0 +1,269 @@
+
+        genome_file = self.param_required('genome_file')
+        gnm = self.param('gnm',self.param_required('genome_file_tmpdir'))
+        msk = self.param_required('msk')
+        rpt = self.param_required('rpt')
+        red_path = self.param_required('red_path')
+        target_db = self.param_required('target_db')
+        self.param('target_db_url',target_db['driver']+'://'+ \
+                                   target_db['user']+':'+ \
+                                   target_db['pass']+'@'+ \
+                                   target_db['host']+':'+ \
+                                   str(target_db['port'])+'/'+ \
+                                   target_db['dbname']+'?'+ \
+                                   'local_infile=1')
+
+        # make sure the genome_file tmpdir 'gnm' exists and copy the genome file into it
+        # in this way we make sure that the only .fa file to be processed is the one we want
+        try:
+            os.makedirs(gnm,exist_ok=True)
+            os.chmod(gnm,0o777)
+        except PermissionError:
+            print('Could not create '+gnm+' directory in "genome_file_tmpdir".')
+            raise
+
+        new_genome_file = gnm+os.path.sep+os.path.basename(genome_file)
+        try:
+            # without read() Python keeps running before the file is written
+            os.popen('cp '+genome_file+' '+new_genome_file).read()
+        except PermissionError:
+            print('Could not copy file '+genome_file+' into directory '+gnm)
+            raise
+
+        # check that the file genome_file exists, it ends with .fa
+        # and there is not any other .fa file within the same directory
+        if sum(map(lambda x: x.endswith('.fa'),os.listdir(gnm))) > 1:
+            raise ValueError('The Red program requires that the directory containing \
+                              the .fa file does not contain any other .fa file. \
+                              The directory '+gnm+' in "gnm"  contains more than one .fa file.')
+        elif not(os.path.isfile(genome_file)):
+            raise ValueError('The file '+genome_file+' in "genome_file" does not end \
+                              with .fa as required by the Red program.')
+
+        # check that the file was copied successfully
+        if not(filecmp.cmp(genome_file,new_genome_file,shallow=False)):
+            raise FileNotFoundError(errno.ENOENT,os.strerror(errno.ENOENT),new_genome_file)
+
+        genome_file = self.param(genome_file,new_genome_file)
+
+        # check that the Red binary exists
+        if not(os.path.isfile(red_path)):
+            raise FileNotFoundError(errno.ENOENT,os.strerror(errno.ENOENT),red_path)
+
+        # check that the target database exists and fetch the seq_region_ids required later
+        if db_utils.database_exists(self.param('target_db_url')):
+            engine = db.create_engine(self.param('target_db_url'))
```

Comment:
What would be the impact of simply trying to connect to the database instead fo testing first if it exists?

Reply:
Oh yes, it's redundant so I've removed the call to database_exists and also removed its package import as it's no longer required. I removed it from the requirements.txt file too.